enzyme/Enzyme/MLIR/Implementations/LinalgAutoDiffOpInterfaceImpl.cpp

```diff
@@ -0,0 +1,237 @@
+          pushCaches.push_back(pushCache);
+          newOpRegion->addArgument(t, loc);
+
+          Value popCache = adjointRegion->insertArgument(
+              numInputsAdjoint + numCaches, t, loc);
+          numCaches++;
+          return std::pair<Value, Value>(pushCache, popCache);
+        };
+
+    gutils->Logic.differentiate(
+        gutils, *linalgOp.getBlock()->getParent(), adjoint.getBodyRegion(),
+        /*parentRegion=*/false, buildFuncReturnOp, hook);
+
+    // TODO add pushCaches to the yield in newOp
+    auto newOpYield = cast<linalg::YieldOp>(
+        cast<linalg::GenericOp>(newOp).getBodyRegion().front().getTerminator());
+    for (auto pc : pushCaches) {
+      newOpYield.getValuesMutable().append(pc);
+    }
+
+    Block *body = &(adjoint.getBodyRegion().front());
+    auto yieldOp = cast<linalg::YieldOp>(body->getTerminator());
+
+    for (auto opOperand : yieldOp.getOperands()) {
+      body->addArgument(opOperand.getType(), opOperand.getLoc());
+    }
+    OpBuilder builderAdd(yieldOp);
+    for (auto it : llvm::enumerate(yieldOp.getOperands())) {
+      Value arg = body->getArgument(outputs.size() + numCaches + it.index());
+      auto diffeType = cast<AutoDiffTypeInterface>(arg.getType());
+      Value grad = diffeType.createAddOp(builderAdd, it.value().getLoc(), arg,
+                                         it.value());
+      yieldOp.setOperand(it.index(), grad);
+    }
+
+    auto indexing_maps = newOp.getIndexingMapsArray();
+    auto indexing_maps_adjoint = adjoint.getIndexingMapsArray();
+    for (int i = 0; i < numCaches; i++) {
+      Value cacheArg = body->getArgument(outputs.size() + i);
+
+      Type ct = cacheArg.getType();
+      Type type = MemRefType::get(ArrayRef(shapes), ct);
+      Value alloc = cacheBuilder.create<memref::AllocOp>(
+          op->getLoc(), type, ValueRange(iterationDomains));
+      Value cache = gutils->initAndPushCache(alloc, cacheBuilder);
+      alloc.getDefiningOp()->setAttr("operand_segment_sizes",
+                                     cacheBuilder.getDenseI32ArrayAttr({1, 0}));
+
+      cast<linalg::GenericOp>(newOp).getOutputsMutable().append(
+          ValueRange({alloc}));
+      indexing_maps.push_back(AffineMap::getMultiDimIdentityMap(
+          iterationDomains.size(), cacheBuilder.getContext()));
+
+      OpBuilder builder2(adjoint);
+      Value retrievedValue = gutils->popCache(cache, builder2);
+      retrievedValue = invertMemref(retrievedValue, builder2, op->getLoc());
+      adjoint.getInputsMutable().append(ValueRange({retrievedValue}));
+      indexing_maps_adjoint.insert(
+          indexing_maps_adjoint.begin() + numInputsAdjoint + i,
+          AffineMap::getMultiDimIdentityMap(iterationDomains.size(),
+                                            builder2.getContext()));
+    }
+    SmallVector<Attribute> indexing_maps_attr;
+    SmallVector<Attribute> indexing_maps_attr_adjoint;
+    for (auto map : indexing_maps) {
```

Comment:
```suggestion
    for (auto &map : indexing_maps) {
```
maybe make this a reference